src/router.js

```diff
@@ -1,72 +1,64 @@
-export function router(options) {
-  return function(emit) {
-    return {
-      state: {
-        router: {}
-      },
-      actions: {
-        router: {
-          set: function(state, actions, data) {
-            return {
-              router: data
-            }
-          },
-          go: function(state, actions, path) {
-            if (location.pathname + location.search !== path) {
-              history.pushState({}, "", path)
-              actions.router.set({
-                path: path
-              })
-            }
-          }
-        }
-      },
-      events: {
-        load: function(state, actions) {
-          addEventListener("popstate", function() {
-            actions.router.set({})
-          })
-        },
-        render: function(state, actions, view) {
-          return view[
-            (state.router.index >= 0
-              ? state
-              : actions.router.set(
-                  emit("route", match(location.pathname, view))
-                )).router.index
-          ][1]
-        }
+import { h } from 'hyperapp'
+
+// router module
+export const router = {
+  actions: {
+    set: function (state, actions, data) {
+      return data
+    },
+    go: function (state, actions, path) {
+      if (location.pathname + location.search !== path) {
+        history.pushState({}, '', path)
+        actions.set({ path: path })
       }
     }
+  },
+  init(state, actions) {
+    window.addEventListener('popstate', function () {
+      actions.set({})
+    })
   }
 }
 
-function match(pathname, routes) {
+// Router component
+export function Router(props, children) {
+  const m = match(location.pathname, children)
+  const i = m.match ? m.index : children.length - 1
+  const v = children[i].component
+  return v(props.state, props.actions)
+}
+
+// Route component
+export function Route(props, children) {
+  return props
+}
+
+function match(pathname, children) {
   var match
   var index
   var params = {}
 
-  for (var i = 0; i < routes.length && !match; i++) {
-    var route = routes[i][0]
+  for (var i = 0; i < children.length && !match; i++) {
+    var route = children[i].path
     var keys = []
     pathname.replace(
       RegExp(
         route === "*"
           ? ".*"
           : "^" +
-            route.replace(/\//g, "\\/").replace(/:([\w]+)/g, function(_, key) {
-              keys.push(key)
-              return "([-\\.%\\w\\(\\)]+)"
-            }) +
-            "/?$",
+          route.replace(/\//g, "\\/").replace(/:([\w]+)/g, function (_, key) {
```

Comment:
@speier @lukejacksonn I know it deserves its own proper issue, but what about parsing the query string too? Just saying.

Reply:
We had this discussion before.. it was dismissed on the grounds that whatever implementation we chose would not be sufficient enough (for all cases).. that said, I still think that a simple query string parser would be useful for most people.

Reply:
Thanks @lukejacksonn, so it's just a matter of deciding how simple is simple enough and lay it out for later implementation. Let's discuss this later then.

Reply:
Yep, exactly! I just tried finding the thread but can't remember where it was discussed 🤔 There was talk of doing what react router does, which IIRC, allows the developer to pass in a query string parsing function; that way we can't be at fault. Perhaps allowing for that, but providing a default parser suitable for >90% of use cases is the way to go.